### PR TITLE
feat(issue-details): Improved UI for viewing object/array values

### LINKED ISSUE #58573: Automatically collapse long context values
### Problem Statement

Doing `Sentry.setContext('title', { value })`, if it goes beyond a certain length, we want `value` to be collapsed in the UI and not truncated. Otherwise a long array could take up a page of space on the screen. Could we have this configurable as an option to `setContext`?

### Solution Brainstorm

_No response_

### Product Area

Issues

┆Issue is synchronized with this [Jira Improvement](https://getsentry.atlassian.net/browse/FEEDBACK-1961) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by malwilley**
Fixes https://github.com/getsentry/sentry/issues/58573, https://github.com/getsentry/sentry/issues/48480

Previously, objects/arrays were only collapsible after they reached maxDefaultDepth. The major change here is that now each level is collapsible

- Each level is now always collapsible. Before, objects/arrays were only collapsible after a certain depth
- Arrays/objects begin collapsed when there are more than 5 items, in addition to when they are at a certain depth
- Changed the appearance and location of the toggle buttons to better match commonly-used json viewers
- Added `n items` preview text for collapsed objects/arrays (which is clickable!)

Before:

CleanShot 2024-03-01 at 10 53 57

After:

CleanShot 2024-03-04 at 09 13 06

And here's an example of the new auto-collapse behavior helping out when context values are very large: 

CleanShot 2024-03-04 at 09 14 12

### CODE REVIEW

**Comment by vuluongj20** [comment]
Can we make the toggles less prominent (perhaps with just the icon and no button formatting) and only appear on hover?

File: static/app/components/structuredEventData/collapsibleValue.tsx [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+export function CollapsibleValue({
+  children,
+  openTag,
+  closeTag,
+  prefix = null,
+  depth,
+  maxDefaultDepth,
+}: CollapsibleValueProps) {
+  const numChildren = Children.count(children);
```
Comment (scttcper) [inline static/app/components/structuredEventData/collapsibleValue.tsx:29]: what

Comment (scttcper) [reply static/app/components/structuredEventData/collapsibleValue.tsx:29]: crazy https://react.dev/reference/react/Children#children-count

**Comment by malwilley** [comment]
@vuluongj20 agree that the toggle buttons make it look a bit busy. Went ahead and removed the background and border, it looks like this (a lot better!):

CleanShot 2024-03-01 at 16 08 21

I tried hiding the buttons until hover, but it left gaps that didn't look very good:

CleanShot 2024-03-01 at 15 48 30

IMO keeping the buttons all the time isn't so bad now that they are less emphasized. What do you think?

**Review by scttcper — APPROVED** [review]

**Comment by vuluongj20** [comment]
Ahh yea I agree with you about the gaps. Let's merge the current version then! 🚀

**Review by vuluongj20 — APPROVED** [review]

**Pull request merged**